src/components/dialog/dialog.ts

```diff
@@ -0,0 +1,270 @@
+import { ifDefined } from 'lit/directives/if-defined.js';
+import { themes } from '../../theming';
+import { watch } from '../common/decorators';
+import { Constructor } from '../common/mixins/constructor';
+import { EventEmitterMixin } from '../common/mixins/event-emitter';
+import { styles } from './themes/light/dialog.base.css';
+import { styles as bootstrap } from './themes/light/dialog.bootstrap.css';
+import { styles as fluent } from './themes/light/dialog.fluent.css';
+import { styles as indigo } from './themes/light/dialog.indigo.css';
+import { styles as material } from './themes/light/dialog.material.css';
+
+let nextId = 0;
+
+export interface IgcDialogEventMap {
+  igcOpening: CustomEvent<void>;
+  igcOpened: CustomEvent<void>;
+  igcClosing: CustomEvent<void>;
+  igcClosed: CustomEvent<void>;
+}
+
+/**
+ * Represents a Dialog component.
+ *
+ * @element igc-dialog
+ *
+ * @fires igcOpening - Emitted just before the dialog is open.
+ * @fires igcOpened - Emitted after the dialog is open.
+ * @fires igcClosing - Emitter just before the dialog is closed.
+ * @fires igcClosed - Emitted after closing the dialog.
+ *
+ * @slot - Renders content inside the default slot.
+ * @slot title - Renders the title of the dialog header.
+ * @slot footer - Renders the dialog footer.
+ *
+ * @csspart base - The base wrapper of the dialog.
+ * @csspart title - The title container.
+ * @csspart footer - The footer container.
+ * @csspart overlay - The overlay.
+ */
+@themes({ bootstrap, material, fluent, indigo })
+export default class IgcDialogComponent extends EventEmitterMixin<
+  IgcDialogEventMap,
+  Constructor<LitElement>
+>(LitElement) {
+  public static readonly tagName = 'igc-dialog';
+  public static styles = [styles];
+
+  private titleId = `title-${nextId++}`;
+  private _ariaLabel!: string;
+  private _ariaLabelledby!: string;
+  private _ariaDescribedby!: string;
+  private _role!: 'dialog' | 'alertdialog';
+
+  @query('[part="base"]')
+  private nativeElement!: any;
+
+  @queryAssignedElements({ slot: 'title' })
+  private _titleElements!: Array<HTMLElement>;
+
+  /** Whether the dialog should be closed when pressing 'ESC' button.  */
+  @property({ type: Boolean, attribute: 'close-on-escape' })
+  public closeOnEscape = true;
+
+  /** Whether the dialog should be closed when clicking outside of it.  */
+  @property({ type: Boolean, attribute: 'close-on-outside-click' })
+  public closeOnOutsideClick = false;
+
+  /** Whether the dialog is opened. */
+  @property({ type: Boolean, reflect: true })
+  public open = false;
+
+  /** Sets the title of the dialog.  */
+  @property({ type: String })
+  public override title!: string;
+
+  /** Sets the role attribute for the control. */
+  @property({ reflect: true })
+  public get role() {
+    return this._role;
+  }
+
+  public set role(value: 'dialog' | 'alertdialog') {
+    const oldVal = this._role;
+    this._role = value;
+    this.removeAttributeAndUpdate('role', 'role', oldVal);
+  }
```

Comment:
Can we set the role attribute in the connectedCallback if not present?
Seems like an overkill?